pkg/storegateway/storegatewaypb/custom.go

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package storegatewaypb
+
+import (
+	"context"
+
+	"github.com/grafana/dskit/grpcutil"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+
+	"github.com/grafana/mimir/pkg/storegateway/storepb"
+)
+
+// customStoreGatewayClient is a custom client which wraps the real gRPC client and re-maps well known
+// gRPC errors into standard golang errors.
+type customStoreGatewayClient struct {
+	wrapped StoreGatewayClient
+}
+
+func NewCustomStoreGatewayClient(cc *grpc.ClientConn) StoreGatewayClient {
+	return &customStoreGatewayClient{
+		wrapped: NewStoreGatewayClient(cc),
+	}
+}
+
+// Series implements StoreGatewayClient.
+func (c *customStoreGatewayClient) Series(ctx context.Context, in *storepb.SeriesRequest, opts ...grpc.CallOption) (StoreGateway_SeriesClient, error) {
+	client, err := c.wrapped.Series(ctx, in, opts...)
+	if err != nil {
+		return client, mapError(err)
+	}
+
+	return &customSeriesClient{client}, nil
+}
+
+// LabelNames implements StoreGatewayClient.
+func (c *customStoreGatewayClient) LabelNames(ctx context.Context, in *storepb.LabelNamesRequest, opts ...grpc.CallOption) (*storepb.LabelNamesResponse, error) {
+	res, err := c.wrapped.LabelNames(ctx, in, opts...)
+	return res, mapError(err)
+}
+
+// LabelValues implements StoreGatewayClient.
+func (c *customStoreGatewayClient) LabelValues(ctx context.Context, in *storepb.LabelValuesRequest, opts ...grpc.CallOption) (*storepb.LabelValuesResponse, error) {
+	res, err := c.wrapped.LabelValues(ctx, in, opts...)
+	return res, mapError(err)
+}
+
+type customSeriesClient struct {
+	StoreGateway_SeriesClient
+}
+
+func (c *customSeriesClient) Recv() (*storepb.SeriesResponse, error) {
```

Comment:
I think we need to do the same kind of mapping for the other client methods (`Send`, `Header`, `CloseSend`...) as well.

Reply:
Agree. The good news is that we need to wrap `grpc.ClientStream` only once and then we can re-use it for other gRPC stream clients: https://github.com/grafana/mimir/pull/6934/commits/bd72c0201572aa5e1ea70fa86ce036d122f32127